blocks/bricks/__init__.py

```diff
@@ -494,45 +495,73 @@ def apply(self, input_):
         return tensor.switch(input_ > 0, input_, 0)
 
 
-class Softmax(Activation):
+class Softmax(Brick):
     @application(inputs=['input_'], outputs=['output'])
     def apply(self, input_):
         return tensor.nnet.softmax(input_)
 
-    @application
-    def categorical_cross_entropy(self, y, x):
-        """Return computationally stable softmax cost.
+    @application(inputs=['input_'], outputs=['output'])
+    def log_probabilities(self, input_):
+        """Normalize log-probabilities.
+
+        Converts unnormalized log-probabilities (exponents of which do not
+        sum to one) into actual log-probabilities (exponents of which sum
+        to one).
+
+        Parameters
+        ----------
+        input_ : :class:`~theano.Variable`
+            A matrix, each row contains unnormalized log-probabilities of a
+            distribution.
+
+        Returns
+        -------
+        output : :class:`~theano.Variable`
+            A matrix with normalized log-probabilities for each
+            distribution from `input_`.
+
+        """
+        shifted = input_ - input_.max(axis=1, keepdims=True)
+        return shifted - tensor.log(
+            tensor.exp(shifted).sum(axis=1, keepdims=True))
+
+    @application(inputs=['y', 'x'], outputs=['output'])
+    def categorical_cross_entropy(self, application_call, y, x):
+        """Computationally stable cross-entropy for pre-softmax values.
 
         Parameters
         ----------
         y : :class:`~tensor.TensorVariable`
             In the case of a matrix argument, each slice along
-            axis represents one distribution. In the vector case, each
+            axis 0 represents one distribution. In the vector case, each
```

Comment:
maybe use terms `rows` and `columns` since you limit this to 2d case. It is much easier to unserstand